File: DemoPhotoSelectionUtility_Objc/DemoPhotoSelectionUtility_Objc/PhotoSelectionUtility.m
//
//  PhotoSelectionUtility.m
//
//  Created by Matthew Sabath
//  MIT License

#import "PhotoSelectionUtility.h"

#import <Photos/Photos.h>

@interface PhotoSelectionUtility () <UIActionSheetDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@property (nonatomic, strong) UIImagePickerController *imagePickerController;
@property (nonatomic, assign) UIViewController *viewController;
@end


@implementation PhotoSelectionUtility

#pragma mark - Post Photo

- (void)submitUserPhotoFromViewController:(UIViewController *)vc
{
    self.viewController = vc;
    
    if ([self systemVersionGreaterThanOrEqual:@"8.0"])
    {
        if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusAuthorized) {
            [self getUserImagePhoto];
        }
        else if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusNotDetermined) {
            [self requestPhotoPermissions];
        }
        else if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusDenied) {
            [self showPhotosDeniedAlert];
        }
    } else {
        // basic handling for iOS versions prior to iOS8
        [self getUserImagePhoto];
    }
}

- (void)getUserImagePhoto
{
    if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera] &&
        [UIImagePickerController isCameraDeviceAvailable:UIImagePickerControllerCameraDeviceRear])
    {
        UIActionSheet *actionSheet = [[UIActionSheet alloc]
                                      initWithTitle:@"Submit Photo"
                                      delegate:self
                                      cancelButtonTitle:@"Cancel"
                                      destructiveButtonTitle:nil
                                      otherButtonTitles:@"Choose From Library",@"Take New Photo", nil];
        
        if (self.viewController.presentedViewController) {
            [actionSheet showInView:self.viewController.presentedViewController.view];
        } else {
            [actionSheet showInView:self.viewController.navigationController.view];
        }
        
    } else {
        [self choosePhotoFromLibrary];
    }
}

- (void)choosePhotoFromLibrary
{
    if (self.imagePickerController == nil) {
        self.imagePickerController = [[UIImagePickerController alloc] init];
    }
    
    self.imagePickerController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    self.imagePickerController.delegate = self;
    [self presentImagePickerController:self.imagePickerController];
}

- (void)takeNewPhoto
{
    if (self.imagePickerController == nil) {
        self.imagePickerController = [[UIImagePickerController alloc] init];
    }
    self.imagePickerController.sourceType = UIImagePickerControllerSourceTypeCamera;
    self.imagePickerController.delegate = self;
    [self presentImagePickerController:self.imagePickerController];
}

- (void)presentImagePickerController:(UIViewController *)pickerController
{
    if (self.viewController.presentedViewController) {
        [self.viewController.presentedViewController presentViewController:pickerController animated:YES completion:^{}];
    } else {
        [self.viewController.navigationController presentViewController:pickerController animated:YES completion:^{}];
    }
}

- (void)requestPhotoPermissions {
    
    [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
        switch (status) {
            case PHAuthorizationStatusAuthorized:
                [self getUserImagePhoto];
                break;
            case PHAuthorizationStatusRestricted:
                break;
            case PHAuthorizationStatusDenied:
                break;
            default:
                break;
        }
    }];
}

- (void)showPhotosDeniedAlert
{
    [[[UIAlertView alloc] initWithTitle:@"Photo Access Denied"
                                message:@"Please go to the Settings app from your device's home screen to re-enable photos."
                               delegate:self
                      cancelButtonTitle:@"OK"
                      otherButtonTitles:nil] show];
}


#pragma mark - ImagePickerDelegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    UIImage *image = [info valueForKey:UIImagePickerControllerOriginalImage];
    picker.delegate = self;
    
    [picker dismissViewControllerAnimated:YES completion:^{
        
        [self.delegate photoSelectionUtilityDidReturnImage:image];
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    [picker dismissViewControllerAnimated:YES completion:^{ }];
}


#pragma mark - Image Manipulation

- (UIImage *)normalizedImage:(UIImage *)image
{
    if (image.imageOrientation == UIImageOrientationUp) return image;
    
    UIGraphicsBeginImageContextWithOptions(image.size, NO, image.scale);
    [image drawInRect:(CGRect){0, 0, image.size}];
    UIImage *normalizedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return normalizedImage;
}


#pragma mark - UIActionSheet Delegate

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if (buttonIndex == 0) {
        [self choosePhotoFromLibrary];
    } else if (buttonIndex == 1) {
        [self performSelectorOnMainThread:@selector(takeNewPhoto) withObject:nil waitUntilDone:NO];
    }
}


#pragma mark - Utility

- (BOOL)systemVersionGreaterThanOrEqual:(NSString *)systemVersion {
    return ([[[UIDevice currentDevice] systemVersion] compare:systemVersion options:NSNumericSearch] != NSOrderedAscending);
}

@end

File: PhotoSelectionUtility/PhotoSelectionUtility.swift
//
//  PhotoSelectionUtility.swift
//  DemoPhotoSelectionUtility_Swift
//
//  Created by Matthew Sabath on 11/10/15.
//  MIT License
//

import Foundation


import Foundation
import Photos
import UIKit

protocol PhotoSelectionUtilityDelegate: class {
    func photoSelectionUtilityDidReturnImage(image: UIImage)
}

class PhotoSelectionUtility: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    weak var delegate: PhotoSelectionUtilityDelegate?
    
    var imagePickerController: UIImagePickerController?
    var viewController: UIViewController?
    
    override init() {
        super.init()
        
    }
    
    func submitUserPhotoFromViewController(vc: UIViewController) {
        
        self.viewController = vc;
        
        if (systemVersionGreaterThanOrEqualToiOS8()) {
            
            let authStatus = PHPhotoLibrary.authorizationStatus()
            
            if (authStatus == .Authorized) {
                getUserImagePhoto()
            } else if (authStatus == .NotDetermined) {
                requestPhotoPermissions()
            } else if (authStatus == .Denied) {
                
            }
        }
    }
    
    func getUserImagePhoto() {
        
        if (UIImagePickerController.isSourceTypeAvailable(.Camera)) {
            let alertController = UIAlertController(title: "Submit Photo", message: nil, preferredStyle: .ActionSheet)
            
            let libraryAction = UIAlertAction(title: "Choose From Library", style: .Default, handler: { _ in
                self.choosePhotoFromLibrary()
            })
            
            let cameraAction = UIAlertAction(title: "Take New Photo", style: .Default, handler: { _ in
                self.takeNewPhoto()
            })
            
            alertController.addAction(libraryAction)
            alertController.addAction(cameraAction)
            
            if let presentedVC = viewController!.presentedViewController {
                presentedVC.presentViewController(alertController, animated: true, completion: nil)
            } else {
                viewController!.presentViewController(alertController, animated: true, completion: nil)
            }
            
        } else {
            choosePhotoFromLibrary()
        }
    }
    
    func takeNewPhoto() {
        imagePickerController = UIImagePickerController()
        imagePickerController!.sourceType = .Camera
        imagePickerController!.delegate = self
        
        presentImagePickerController(imagePickerController!)
    }
    
    func choosePhotoFromLibrary() {
        imagePickerController = UIImagePickerController()
        imagePickerController!.sourceType = .PhotoLibrary
        imagePickerController!.delegate = self
        
        presentImagePickerController(imagePickerController!)
    }
    
    func presentImagePickerController(pickerController: UIViewController) {
        if let _ = viewController!.presentedViewController {
            viewController!.presentedViewController!.presentViewController(pickerController, animated: true, completion:nil)
        } else {
            viewController!.navigationController?.presentViewController(pickerController, animated: true, completion:nil)
        }
    }
    
    func requestPhotoPermissions() {
        PHPhotoLibrary.requestAuthorization { (status) -> Void in
            
            switch (status) {
            case .Authorized:
                self.getUserImagePhoto()
                break;
                
            case .Restricted:
                // handle
                break;
                
            case .Denied:
                // handle
                break;
                
            default:
                break;
            }
        }
    }
    
    
    // MARK: ImagePickerDelegate
    
    func imagePickerController(picker: UIImagePickerController, didFinishPickingImage image: UIImage, editingInfo: [String : AnyObject]?) {
        picker.delegate = self;
        picker.dismissViewControllerAnimated(true, completion: { _ in
            
            self.delegate?.photoSelectionUtilityDidReturnImage(image)
        })
    }
    
    func imagePickerControllerDidCancel(picker: UIImagePickerController) {
        picker.dismissViewControllerAnimated(true, completion: nil)
    }
    
    // MARK: Alerts
    
    func throwPhotosDeniedAlert() {
        let alertController = UIAlertController(title: "Photo Access Denied", message: "Please go to the Settings app from your device's home screen to re-enable photos.", preferredStyle: .Alert)
        
        let okAction = UIAlertAction(title: "Ok", style: .Default, handler: { _ in
            self.viewController!.dismissViewControllerAnimated(true, completion: nil)
        })
        
        alertController.addAction(okAction)
        viewController!.presentViewController(alertController, animated: true, completion: nil)
    }
    
    
    // MARK: Utility
    
    func systemVersionGreaterThanOrEqualToiOS8() -> Bool {
        return NSProcessInfo().isOperatingSystemAtLeastVersion(NSOperatingSystemVersion(majorVersion: 8, minorVersion: 0, patchVersion: 0))
    }
}